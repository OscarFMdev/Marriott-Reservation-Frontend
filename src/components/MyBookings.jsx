import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { ProgressSpinner } from 'primereact/progressspinner';
import { fetchBookings, deleteBooking, popBook } from '../redux/reducer/reservation/reservationSlice';
import { setMessageEmpty } from '../redux/reducer/Rooms/roomSlice';
import mybookings from './componentsCss/TableStyles.module.css';

const MyBookings = () => {
  const { status, message, bookings } = useSelector((state) => state.reservations);
  const { user } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const handleCancelReservation = (id) => {
    const deleteBook = {
      id,
      userId: user.id,
    };
    dispatch(deleteBooking(deleteBook));
    dispatch(popBook(id));
  };

  useEffect(() => {
    dispatch(fetchBookings(user.id));
  }, []);

  const handleMessage = () => {
    setTimeout(() => {
      if (message === 'Room Created Successfully') dispatch(setMessageEmpty(''));
    }, 3000);
  };

  useEffect(() => {
    handleMessage();
  }, [message]);

  if (status === 'loading') {
    return <ProgressSpinner className="progressSpinner" />;
  }

  return (
    <main className={mybookings.container}>
      <h1>My Bookings</h1>
      {message && (
        <p className={message ? `${mybookings.success} ${mybookings.fadeIn}` : mybookings.fadeOut}>
          {message}
        </p>
      )}
      <table className="">
        <thead>
          <tr>
            <th>Room Name</th>
            <th>Check-in Date</th>
            <th>Check-out Date</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {bookings.length === 0 ? (
            <tr>
              <td colSpan="4">You have no bookings</td>
            </tr>
          ) : (
            bookings.map(
              ({
                id: bookingId,
                start_date: startDate,
                end_date: endDate,
                room: {
                  name,
                },
              }) => (
                <tr key={bookingId} className="hello">
                  <>
                    <td>{name}</td>
                    <td>{startDate}</td>
                    <td>{endDate}</td>
                  </>
                  <td>
                    <button
                      type="button"
                      onClick={() => handleCancelReservation(bookingId)}
                      className={mybookings.deleteCancelBtn}
                    >
                      Cancel
                    </button>
                  </td>
                </tr>
              ),
            )
          )}
        </tbody>
      </table>

    </main>
  );
};

export default MyBookings;
