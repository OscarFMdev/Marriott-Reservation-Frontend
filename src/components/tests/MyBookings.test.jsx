import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import MyBookings from '../MyBookings';
import store from '../../redux/configureStore';

describe('It displays the Booking page title', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <Router>
          <MyBookings />
        </Router>
      </Provider>,
    );
  });
  test('My Booking title', () => {
    expect(screen.getByText('My Bookings')).toBeDefined();
  });
});

test('My Bookings page renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <MyBookings />
      </Router>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
