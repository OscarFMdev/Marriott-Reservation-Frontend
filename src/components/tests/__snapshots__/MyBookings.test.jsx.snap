// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`My Bookings page renders correctly 1`] = `
<main
  className="_container_ca1eb0"
>
  <h1>
    My Bookings
  </h1>
  <table
    className=""
  >
    <thead>
      <tr>
        <th>
          Room Name
        </th>
        <th>
          Check-in Date
        </th>
        <th>
          Check-out Date
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan="4"
        >
          You have no bookings
        </td>
      </tr>
    </tbody>
  </table>
</main>
`;
